### PR TITLE
Add type hints to room member handlers

### PR DESCRIPTION
**Pull request by clokep**
The promised follow-up to #7507 with type hints and updated comments.

Note that some of the return types are...less than clear here, so checking them over would be helpful!

### CODE REVIEW

File: synapse/handlers/room_member.py [hunk 1]
```diff
@@ -484,19 +486,20 @@ async def _update_membership(
...
 
         Args:
-            old_room_id (str): The ID of the old room
+            old_room_id: The ID of the old room
 
-            room_id (str): The ID of the new room
+            room_id: The ID of the new room
 
         Returns:
             Deferred
```
Comment (babolivier) [inline synapse/handlers/room_member.py:505]: I don't think that's still true.

File: synapse/handlers/room_member.py [hunk 2]
```diff
@@ -566,7 +568,13 @@ async def copy_user_state_on_room_upgrade(self, old_room_id, new_room_id, user_i
                 )
                 continue
 
-    async def send_membership_event(self, requester, event, context, ratelimit=True):
+    async def send_membership_event(
+        self,
+        requester: Requester,
+        event: EventBase,
+        context: EventContext,
+        ratelimit: bool = True,
```
Comment (babolivier) [inline synapse/handlers/room_member.py:576]: You probably want to remove the types from the docstring of that function (especially as the types listed there seem outdated).

File: synapse/handlers/room_member.py [hunk 3]
```diff
@@ -887,21 +904,26 @@ async def _is_remote_room_too_complex(self, room_id, remote_room_hosts):
...
-    async def _is_local_room_too_complex(self, room_id):
+    async def _is_local_room_too_complex(self, room_id: str) -> bool:
         """
         Check if the complexity of a local room is too great.
 
         Args:
-            room_id (str)
-
-        Returns: bool
+            room_id
```
Comment (babolivier) [inline synapse/handlers/room_member.py:912]: This line seems useless now that it doesn't even have the type. Could you please remove it or add a documentation about this parameter?

File: synapse/handlers/room_member.py (hunk 1 above)
Comment (clokep) [reply synapse/handlers/room_member.py:505]: Arg I missed one! Good catch. :)

**Pull request merged**